### PR TITLE
Fix pathfinding

### PR DESCRIPTION
**Pull request by glebm**
The previous implementation didn't behave quite like A-* is supposed to.

After trying to figure out what's causing it and giving up, I've reimplemented it in a straightforward manner.
Now it seems to work a lot better.

Also increases maximum player path length to 100 steps.
We still only store the first 25 steps in the save file for vanilla compatibility.

Why draft?
1. Need to check how much smarter monsters are with this.
2. New demo needs to be recorded.

Benchmarking shows that the new algorithm is about twice as fast:

```
Comparing build-reld-path-bench/path_benchmark to build-reld/path_benchmark
Benchmark                              Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------
BM_SinglePath_pvalue                 0.0002          0.0002      U Test, Repetitions: 10 vs 10
BM_SinglePath_mean                  -0.3270         -0.3271          2917          1963          2917          1963
BM_SinglePath_median                -0.3247         -0.3247          2907          1963          2907          1963
BM_SinglePath_stddev                -0.8035         -0.8067            15             3            15             3
BM_SinglePath_cv                    -0.7080         -0.7127             0             0             0             0
BM_Bridges_pvalue                    0.0002          0.0002      U Test, Repetitions: 10 vs 10
BM_Bridges_mean                     -0.7641         -0.7641         59742         14095         59736         14094
BM_Bridges_median                   -0.7617         -0.7618         59083         14077         59077         14074
BM_Bridges_stddev                   -0.9810         -0.9810          2719            52          2719            52
BM_Bridges_cv                       -0.9194         -0.9195             0             0             0             0
BM_NoPath_pvalue                     0.0002          0.0002      U Test, Repetitions: 10 vs 10
BM_NoPath_mean                      -0.8406         -0.8406        127807         20371        127792         20368
BM_NoPath_median                    -0.8406         -0.8406        127556         20330        127540         20328
BM_NoPath_stddev                    -0.9138         -0.9141          1109            96          1111            95
BM_NoPath_cv                        -0.4592         -0.4609             0             0             0             0
OVERALL_GEOMEAN                     -0.4405         -0.4405             0             0             0             0
```

If we go ahead with this fix, we can consider increasing the max path length for the player in a follow-up PR.

Here is what a somewhat long path looks like:

path

We may want to tune it further to prefer diagonal steps first more reliably.

### CODE REVIEW

File: test/path_test.cpp [hunk 1]
```diff
@@ -121,33 +123,33 @@ void CheckPath(Point startPosition, Point destinationPosition, std::vector<int8_
...
+// 	CheckPath({ 8, 8 }, { 8, 6 }, { 1, 1 });
+// 	CheckPath({ 8, 8 }, { 6, 8 }, { 2, 2 });
+// 	CheckPath({ 8, 8 }, { 10, 8 }, { 3, 3 });
+// 	CheckPath({ 8, 8 }, { 8, 10 }, { 4, 4 });
+
+// 	// Otherwise pathing biases along diagonals and the diagonal steps will always be first
+// 	CheckPath({ 8, 8 }, { 5, 6 }, { 5, 5, 2 });
+// 	CheckPath({ 8, 8 }, { 4, 4 }, { 5, 5, 5, 5 });
+// 	CheckPath({ 8, 8 }, { 12, 20 }, { 7, 7, 7, 7, 4, 4, 4, 4, 4, 4, 4, 4 });
+// }
```
Comment (ephphatha) [inline test/path_test.cpp:140]: The intent for these tests is to ensure the pathfinding algo is consistent/deterministic. If you get this implementation performant enough that you're happy with it and it gives a consistent path between two points on a fixed grid I'm happy to have the list of nodes here changed to match whatever the output ends up giving.

File: test/path_test.cpp [hunk 2]
```diff
@@ -121,33 +123,33 @@ void CheckPath(Point startPosition, Point destinationPosition, std::vector<int8_
...
-	CheckPath({ 56, 56 }, { 0, 0 }, {});
+	// CheckPath({ 56, 56 }, { 0, 0 }, {});
 
 	// Longest possible path is currently 24 steps meaning tiles 24 units away are reachable
 	Point startingPosition { 56, 56 };
 	CheckPath(startingPosition, startingPosition + Displacement { 24, 24 }, std::vector<int8_t>(24, 7));
 
 	// But trying to navigate 25 units fails
-	CheckPath(startingPosition, startingPosition + Displacement { 25, 25 }, {});
+	// CheckPath(startingPosition, startingPosition + Displacement { 25, 25 }, {});
```
Comment (ephphatha) [inline test/path_test.cpp:152]: And these two are purely to document the edge case for the current 24/25 step limit. If that is lifted happy to have this updated to mark the new limit

File: test/path_test.cpp (hunk 1 above)
Comment (glebm) [reply test/path_test.cpp:140]: Oh yeah I didn't mean to comment this out

File: test/path_test.cpp (hunk 2 above)
Comment (glebm) [reply test/path_test.cpp]: I didn't mean to push the test change, reverted

**Comment by kphoenix137** [comment]
Does this PR also fix https://github.com/diasurgical/devilutionX/issues/6103 ?

**Comment by glebm** [comment]
Probably not, this fixes the pathfinding algorithm but doesn't change how stuns are handled in-game.

**Comment by glebm** [comment]
@ephphatha @qndel @AJenbo Ready for review and playtesting!

**Review by AJenbo — APPROVED** [review]
